sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,968 @@
+        Coder<V> valueCoder) {
+      this(conf, keyCoder, valueCoder, null, null, null, null);
+    }
+
+    public HadoopInputFormatBoundedSource(
+        SerializableConfiguration conf,
+        Coder<K> keyCoder,
+        Coder<V> valueCoder,
+        SimpleFunction<?, K> keyTranslationFunction,
+        SimpleFunction<?, V> valueTranslationFunction,
+        SerializableSplit inputSplit) {
+      this(conf, keyCoder, valueCoder, keyTranslationFunction, valueTranslationFunction, null,
+          inputSplit);
+    }
+
+    public HadoopInputFormatBoundedSource(
+        SerializableConfiguration conf,
+        Coder<K> keyCoder,
+        Coder<V> valueCoder,
+        SimpleFunction<?, K> keyTranslationFunction,
+        SimpleFunction<?, V> valueTranslationFunction,
+        InputFormat<?, ?> inputFormatObj,
+        SerializableSplit inputSplit) {
+      this.conf = conf;
+      this.inputSplit = inputSplit;
+      this.keyCoder = keyCoder;
+      this.valueCoder = valueCoder;
+      this.keyTranslationFunction = keyTranslationFunction;
+      this.valueTranslationFunction = valueTranslationFunction;
+      this.inputFormatObj = inputFormatObj;
+    }
+
+    public SerializableConfiguration getConfiguration() {
+      return conf;
+    }
+
+    @Override
+    public void validate() {
+      checkNotNull(conf, HadoopInputFormatIOContants.MISSING_CONFIGURATION_SOURCE_ERROR_MSG);
+      checkNotNull(keyCoder, HadoopInputFormatIOContants.MISSING_KEY_CODER_SOURCE_ERROR_MSG);
+      checkNotNull(valueCoder, HadoopInputFormatIOContants.MISSING_VALUE_CODER_SOURCE_ERROR_MSG);
+    }
+
+    @Override
+    public List<BoundedSource<KV<K, V>>> splitIntoBundles(long desiredBundleSizeBytes,
+        PipelineOptions options) throws Exception {
+      if (inputSplit == null) {
+        if (inputSplits == null) {
+          computeSplits();
+        }
+        LOG.info("Generated {} splits each of size {} ", inputSplits.size(),
+            inputSplits.get(0).getSplit().getLength());
+        return Lists.transform(inputSplits,
+            new Function<SerializableSplit, BoundedSource<KV<K, V>>>() {
+              @Override
+              public BoundedSource<KV<K, V>> apply(SerializableSplit serializableInputSplit) {
+                HadoopInputFormatBoundedSource<K, V> hifBoundedSource =
+                    new HadoopInputFormatBoundedSource<K, V>(conf, keyCoder, valueCoder,
+                        keyTranslationFunction, valueTranslationFunction, inputFormatObj,
+                        serializableInputSplit);
+                return hifBoundedSource;
+              }
+            });
+      } else {
+        LOG.info("Not splitting source {} because source is already split.", this);
+        return ImmutableList.of((BoundedSource<KV<K, V>>) this);
+      }
+    }
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions po) throws Exception {
+      if (inputSplit == null) {
+        if (inputSplits == null) {
```

Comment:
Changed, extracted a method computeSplitsIfNecessary() to do the needful as per your suggestion.